Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -0,0 +1,393 @@
+        let row = indexPath.row
+        let bounds = self.collectionView!.bounds
+
+        let verticalItemsCount = maxVerticalItemsCount()
+        let horizontalItemsCount = maxHorizontalItemsCount()
+
+
+        let itemsPerPage = verticalItemsCount * horizontalItemsCount
+
+        let columnPosition = row % horizontalItemsCount
+        let rowPosition = (row / horizontalItemsCount) % verticalItemsCount
+        let itemPage = Int(floor(Double(row)/Double(itemsPerPage)))
+
+        let attr = UICollectionViewLayoutAttributes(forCellWithIndexPath: indexPath)
+
+        var frame = CGRectZero
+        frame.origin.x = CGFloat(itemPage) * bounds.size.width + CGFloat(columnPosition) * (itemSize.width + insets.left) + insets.left
+        frame.origin.y = CGFloat(rowPosition) * (itemSize.height + insets.top) + insets.top
+        frame.size = itemSize
+        attr.frame = frame
+        
+        return attr
+    }
+}
+
+/*
+    Defines the number of items to show in topsites for different size classes.
+*/
+struct ASTopSiteSourceUX {
+    static let verticalItemsForTraitSizes = [UIUserInterfaceSizeClass.Compact : 1, UIUserInterfaceSizeClass.Regular : 2]
+    static let horizontalItemsForTraitSizes = [UIUserInterfaceSizeClass.Compact : 3, UIUserInterfaceSizeClass.Regular : 5]
+    static let maxNumberOfPages = 2
+    static let CellIdentifier = "TopSiteCell"
+}
+
+protocol ASHorizontalScrollDelegate {
+    func numberOfVerticalItems(estimatedItems: Int) -> Int
+    func numberOfHorizontalItems(estimatedItems: Int) -> Int
+}
+
+/*
+ This Delegate/DataSource is used to manage the ASHorizontalScrollCell's UICollectionView. 
+ This is left generic enough for it to be re used for other parts of Activity Stream.
+ */
+
+class ASHorizontalScrollSource: NSObject, UICollectionViewDelegate, UICollectionViewDataSource, ASHorizontalScrollDelegate {
+
+    var content: [TopSiteItem] = []
+    var urlPressedHandler: ((NSURL) -> Void)?
+    var pageChangedHandler: ((CGFloat) -> Void)?
+
+    // The current traits that define the parent ViewController. Used to determine how many rows/columns should be created.
+    var currentTraits: UITraitCollection?
+
+    /*
+        Size classes define how many items to show per row/column.
+    */
+    func numberOfVerticalItems(estimatedItems: Int) -> Int {
+        if let traits = currentTraits {
+            return ASTopSiteSourceUX.verticalItemsForTraitSizes[traits.verticalSizeClass]!
+        } else {
+            return estimatedItems
+        }
+    }
+
+    func numberOfHorizontalItems(estimatedItems: Int) -> Int {
+        if let traits = currentTraits {
+            // An iPhone 5 in both landscape/portrait is considered compactWidth which means we need to let the layout determine how many items to show based on actual width.
+            return max(ASTopSiteSourceUX.horizontalItemsForTraitSizes[traits.horizontalSizeClass]!, estimatedItems)
```

Comment:
Shouldn't normally be an issue but force unwrapping this might becoming trouble if Apple introduces some weird size class in the future. I think we can move things around to clean this up.

Since the traitCollection is the main input into determining a set of values, I would toss `ASTopSiteSourceUX` and create  a data structure to hold the values like this:

```swift
struct ASHorizontalScrollSourceAttributes {
    let verticalItemCount: Int
    let horizontalItemCount: Int
    let maxPageCount: Int
}
```

.. and include a static constructor that outputs what you need for a trait collection:

```swift
static func attributesForTraitCollection(traitCollection: UITraitCollection) -> ASHorizontalScrollSourceAttributes? {
    ...
}
```

Lastly, you can then be more declarative by create private local instances of the different attributes you'd want:

```swift
private let wideAttributes = ASHorizontalScrollSourceAttributes(
    verticalItemCount: 1,
    horizontalItemCount: 5,
    maxPageCount: 3
)
```

You can might be able to even expose the counts as computed properties that wrap an instance of the attributes struct instead of having them as methods in the delegate.